pkg/api/vlabs/validate.go

```diff
@@ -617,6 +617,16 @@ func (a *Properties) validateAddons() error {
 				if to.Bool(addon.Enabled) && a.HasCoreOS() {
 					return errors.New("flexvolume add-ons not currently supported on coreos distro. Please use Ubuntu")
 				}
+			case "kube-dns":
+				if to.Bool(addon.Enabled) {
+					maxVersion, err := semver.Make("1.12.0-alpha.1")
+					if err != nil {
+						return errors.New("could not validate version")
+					}
+					if !sv.LT(maxVersion) {
```

Comment:
how does this work on upgrades to 1.12 if kube-dns was previously enabled?

Reply:
The current code suggests there's no way to get kube-dns (via aks-engine) w/ 1.12 or greater at present, so this validation is meant to carry over the current "non-user-configurable" functionality into the "user-configurable" area. Check out the delta in the `artifacts.go` implementation to validate whether or not I'm correct about that.

Reply:
the previous code didn't use the isEnabled apimodel value to determine if the addon was enabled so that would work but the way I understand your changes it the behavior is now to use `isEnabled` as source of truth which means it _is_ configurable now and if it is enabled previously upgrade won't change the value since the apimodel values don't get overwritten unless the addon is missing. I believe this breaks back compat.

Reply:
also if we don't want to make it user-configurable we also need to validate that it's enabled for < 1.12

Reply:
The current code (before this PR) as I read it didn't expose `kube-dns` as an api model addon at all. It was implemented as a k8s addon, but it wasn't implemented as an `aks-engine` addon.

Reply:
we could run two tests to prove this works:
1) create 1.11 cluster from previous aks-engine (master branch for eg.) and upgrade to 1.12 with this branch
2) create 1.11 cluster with this branch, upgrade to 1.12 with this branch
3) create 1.11 cluster with this branch with apimodel that has "kube-dns" addon set to `Enabled: false`